chore(#1658): add openmrs and ltfu docs (#1657)
medic/cht-docs: content/en/building/guides/interoperability/openmrs.md

Reviewer: A similar [workflow](https://github.com/medic/cht-interoperability/blob/main/WORKFLOW.md) testing doc is an excellent complement to this documentation. And the corresponding [Postman collection for local testing](https://github.com/medic/cht-interoperability/tree/main/docs/local-test)

Integrating CHT apps with OpenMRS can be achieved using the CHT's [Interoperability Tools]({{< ref "building/guides/interoperability/openhim" >}}) and following the guidance in the [Building interoperability with FHIR APIs documentation]({{< ref "building/guides/interoperability/cht-config" >}}). 

## Overview

The CHT's interoperability tools support interoperability with OpenMRS in a variety of ways:

1. Sending patient and patient contact data.
1. Sending reports data (encounters and observations).
1. Receiving reports data (encounters and observations) from OpenMRS.

The steps to create an OpenMRS interoperability workflow are:

1. Profile the workflow in terms of what data needs to be exchanged between OpenMRS and the CHT application.

> 1. Set up a test environment including an OpenMRS instance, a CHT instance, and OpenHIM. The [interoperability project](https://github.com/medic/cht-interoperability) has docker compose files so that you can set this up locally by running `startup.sh up-openmrs` in the interoperability project.

Reviewer 2: ```suggestion
1. Set up a test environment including an OpenMRS instance, a CHT instance, and OpenHIM. The [interoperability project](https://github.com/medic/cht-interoperability) has docker compose files so that you can set this up locally by running `./startup.sh up-openmrs` in the interoperability project.
```

> 1. Set up a test environment including an OpenMRS instance, a CHT instance, and OpenHIM. The [interoperability project](https://github.com/medic/cht-interoperability) has docker compose files so that you can set this up locally by running `./startup.sh up-openmrs` in the interoperability project.

Author: added cht-config resources created by `./startup.sh up-openmrs` in the OpenHIM resources section below

## Profiling

The first step is to profile the workflow.

1. Which patients should be sent to OpenMRS, and how is a patient defined in the CHT application?
2. Which forms should be sent to OpenMRS?
3. Which fields on those forms should be sent to OpenMRS, and which concepts do they map to?
5. Which forms or other data from OpenMRS should be sent back to CHT?

## Configuring CHT And OpenMRS

Depending on what data needs to be exchanged, [outbound push]({{< ref "building/reference/app-settings/outbound" >}}) configurations and JSON forms need to be added to CHT.

### Sending patients CHT->OpenMRS

When sending patient data to OpenMRS, configure an outbound push mapping as described in the [CHT FHIR config documentation]({{< ref "building/guides/interoperability/cht-config#outbound-patients" >}}).
Patients synced to OpenMRS will have two new [identifier types](https://guide.openmrs.org/getting-started/openmrs-information-model/#patient-identifier): `CHT Document Id`, the uuid of the document that is sent, and `CHT Patient ID`, if there is a `patient_id` field on the patient document.
These identifier types are created automatically when the OpenMRS Channel is registered.

> After setting up the outbound push config, test that it works in the test environment by creating a patient in the CHT application.

Author: added cht-config resources created by `./startup.sh up-openmrs` in the OpenHIM resources section below

```bash
curl -X GET localhost:8090/openmrs/ws/fhir2/R4/Patient/?identifier=[identifier] -H "Authorization: Basic $(echo -n admin:Admin123 | base64)"
```

This sequence diagrams shows the entire flow including the OpenMRS Mediator and the intermediate FHIR Server.
![](cht-outgoing-patients.png)

### Sending forms CHT->OpenMRS

Any data sent from CHT to OpenMRS needs to map to a [concept](https://wiki.openmrs.org/display/docs/Concept+Dictionary+Basics) in OpenMRS. Each concept has a code which will be used to identify the concept in the CHT Application, the FHIR Server, and OpenMRS.
For any fields to send to OpenMRS, first find or create matching concepts in OpenMRS.
Then, using the appropriate codes, configure an outbound push as described in the [CHT FHIR config documentation]({{< ref "building/guides/interoperability/cht-config#outbound-reports" >}}).

In OpenMRS, all form submissions are represented as `Home Visit` encounter types, with a `Visit Note` encounter.
Any fields in the outbound push config are converted to FHIR observations, which are linked to the `Visit Note`.

> After setting up the outbound push, test that it works in the test environment by submitting a report to the form in the CHT application.

Author: added cht-config resources created by `./startup.sh up-openmrs` in the OpenHIM resources section below

      "reported_date": 1730355950419,
      "observations": [
        {
          "code": "5090AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
          "valueQuantity": {
            "value": 175,
            "unit": "cm"
          },
          "label": "Height in cm"
        },
        {
          "code": "5089AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA",
          "valueQuantity": {
            "value": 60,
            "unit": "kg",
          },
          "label": "Weight in kg"
        }
      ]
    }
    ```
1. A PUT request to the FHIR server to create an encounter, and one request for each observation contained in that encounter.
    ![](openhim-put-encounter.png)
1. After a maximum of 1 minute, a polling request from the OpenMRS mediator, and related GET requests to the FHIR server and OpenMRS. The results from the FHIR server should include an encounter to represent the form submission, and observations for any fields that were mapped to concepts.
    ![](openmrs-sync-encounter.png)
1. Requests to OpenMRS to create a `Home Visit` encounter, a `Visit Note` encounter, and one request for each observation.
    ![](openmrs-post-observation.png)

> 1. A request to the FHIR server updating the encounter with the corresponding id from OpenMRS.

Reviewer: Ok, after creating the CHT report, I see only two POST request and then all GET. 
This is how the transaction logs looks like:
![image](https://github.com/user-attachments/assets/24ce5c6a-4075-43be-8e57-188c3499f7c8)

Reviewer: Resolving this conversation as I created [this](Failure to Sync Encounters and Observations from CHT to OpenMRS via Mediator](https://github.com/medic/cht-interoperability/issues/157#top) issue to follow up

If all the above look OK, you should now be able to see the encounter in OpenMRS.

```bash
curl -X GET localhost:8090/openmrs/ws/fhir2/R4/Encounter/?identifier=[identifier] -H "Authorization: Basic $(echo -n admin:Admin123 | base64)"
```

This sequence diagrams shows the entire flow including the OpenMRS Mediator and the intermediate FHIR Server.
![](cht-form-submission.png)

### Sending forms OpenMRS->CHT

> When sending form data to CHT, first find, create, or import the concepts in OpenMRS, and create the forms in OpenMRS.

Reviewer: If there are instructions on how to do this, that would be great. Maybe a link to the OpenMRS page?

> Then create a form in CHT to receive the reports [as described in the CHT interop config documentation]({{< ref "building/guides/interoperability/cht-config#outbound-reports" >}}); the codes from OpenMRS are the field names, and labels can be added for human readability.

Reviewer: If this is pre-configured in the cht-interoperability repository, please include a notice here.

---
title: "OpenMRS Interoperability"

> linkTitle: "OpenMRS Interoperability"

Reviewer 2: ```suggestion
linkTitle: "OpenMRS"
```

weight:
description: >

>    Exchange data with systems based on the OpenMRS using FHIR APIs

Reviewer 3: ```suggestion
   Exchange data with systems based on OpenMRS using FHIR APIs
```

keywords: openmrs interoperability
relatedContent: >

>   building/guides/interoperability/cht_config

Reviewer 2: ```suggestion
  building/guides/interoperability/cht-config
```

  building/concepts/interoperability
  building/guides/interoperability/openhim
---

[OpenMRS](https://openmrs.org) is an open-source electronic medical record system used in dozens of countries. Integrating CHT apps with OpenMRS can open up opportunities to improve health outcomes for patients by promoting better coordination of care. For example, referrals by CHWs in the community can be sent electronically to health facilities using OpenMRS so that nurses and clinicians can prepare for their visit and have full access to the assessment done by a CHW.

> Integrating CHT apps with OpenMRS can be achieved using the CHT's [Interoperability Tools]({{< ref "building/guides/interoperability/openhim" >}}) and following the guidance in the [Building interoperability with FHIR APIs documentation]({{< ref "building/guides/interoperability/cht_config" >}}). 

Reviewer 2: ```suggestion
Integrating CHT apps with OpenMRS can be achieved using the CHT's [Interoperability Tools]({{< ref "building/guides/interoperability/openhim" >}}) and following the guidance in the [Building interoperability with FHIR APIs documentation]({{< ref "building/guides/interoperability/cht-config" >}}). 
```

## Overview

The CHT's interoperability tools support interoperability with OpenMRS in a variety of ways:

1. Sending patient and patient contact data.
2. Sending reports data (encounters and observations).
3. Receiving patient and patient contact data from OpenMRS.
4. Receiving reports data (encounters and observations) from OpenMRS.

> The steps to create an OpenMRS interoperability workflow are:

Author: do you have any specific suggestions?

1. Profile the workflow in terms of what data needs to be exchanged between OpenMRS and the CHT application.

> 1. Set up a test environment including an OpenMRS instance, a CHT instance, and OpenHIM. The [interoperability project](https://github.com/medic/cht-interoperability) has docker compose files so that you can set this up locally by running `startup.sh up-test` in the interoperability project.

Author: changed to `up-openmrs`

> 1. Create or find concepts in OpenMRS that represent any data that needs to be exchanged.

Author: linked to OpenMRS documentation

1. Configure outbound push and forms in the CHT application to match the interoperability workflow.
1. Test and debug any configuration issues in the test environment.
1. Once the configuration is confirmed to be working as expected, start deploying to production by uploading the CHT configuration to production.
1. Add any OpenMRS concepts or forms to production.

> 1. Set OpenMRS credentials and CHT credentials in the interoperability project, and start it in a production deployment.

Author: linked to openhim setup section

This intermediate storage is used to link patients and encounters from OpenMRS and CHT with identifiers from both CHT and OpenMRS and to store additional FHIR metadata.
* An OpenMRS mediator polls the OpenMRS FHIR API periodically (the default is every minute) and synchronizes resources between the FHIR server and OpenMRS.
Synchronizing involves comparing the resources from the FHIR server and OpenMRS and forwarding any new or updated data from one to the other.
This is used both to create CHT data in OpenMRS and to create OpenMRS data in a CHT application.

## Profiling

The first step is to profile the workflow.

1. Which patients should be sent to OpenMRS, and how is a patient defined in the CHT application?
2. Which forms should be sent to OpenMRS?
3. Which fields on those forms should be sent to OpenMRS, and which concepts do they map to?
4. Should patients from OpenMRS be sent back to CHT? If so, what will be used to assign them to CHWs or CHW areas?
5. Which forms or other data from OpenMRS should be sent back to CHT?

## Configuring CHT And OpenMRS

> Depending on what data needs to be exchanged, outbound push configurations and JSON forms need to be added to CHT.

Reviewer 3: Link to outbound push documentation here.

### Sending patients CHT->OpenMRS

When sending patient data to OpenMRS, configure an outbound push mapping as described in the [CHT FHIR config documentation]({{< ref "building/guides/interoperability/cht_config#outbound-patients" >}}).
Patients synced to OpenMRS will have two new [identifier types](https://guide.openmrs.org/getting-started/openmrs-information-model/#patient-identifier): `CHT Document Id`, the uuid of the document that is sent, and `CHT Patient ID`, if there is a `patient_id` field on the patient document.
These identifier types are created automatically when the OpenMRS Channel is registered.

> After setting up the putbound push, test that it works in the test environment by creating a patient in the CHT application.

Reviewer 3: ```suggestion
After setting up the outbound push config, test that it works in the test environment by creating a patient in the CHT application.
```

        },
        "encounter": {
          "reference": "Encounter/8d5e7ddb-1669-4b34-9804-dfae5a15e6a2",
          "type": "Encounter"
        },
        "effectiveDateTime": "2024-11-08T11:07:57+00:00",
        "issued": "2024-11-08T11:07:57.000+00:00",
        "valueQuantity": {
          "value": 60,
          "unit": "kg"
        }
      }
    ]
    ```
1. A request to the records API in the CHT application with any observations mapped to codes.

If all the above look OK, you should now be able to see the data submitted in the CHT.

This sequence diagrams shows the entire flow including the OpenMRS Mediator and the intermediate FHIR Server.
![](cht-incoming-forms.png)

### Sending patients OpenMRS->CHT

Creating patients in a CHT application from patients registered in OpenMRS is possible, but requires some additional configuration to assign patients to a CHW or CHW Area.
All patients in CHT applications require a parent, which is assumed to be a CHW area or equivalent and defines which CHW the patient is assigned to.
For interoperability with OpenMRS, this means that patients created by OpenMRS must be matched to CHT locations.

> A default implementation is provided which uses the [OpenMRS address add on](https://addons.openmrs.org/show/org.openmrs.module.addresshierarchy) to match locations in OpenMRS to contacts in CHT.

Author: Its not included in the openmrs docker compose that comes with cht-interoperability. 
removed all references to incoming patients

This can be complicated to set up and maintain; an alternative is to customize the OpenMRS mediator for specific applications.
If using the default implementation
1. Install the address hierarchy add-on.
2. Download the contact hierarchy from the CHT application.
3. Upload these contacts to OpenMRS: address 5 should be a CHW area, or the direct parent of the patient. If address 5 is not specified, the mediator will use address 4. 

> 4. The mediator will attempt to find CHT locations by using a place id formatted like `[12345]` at the end of the address string. `place_id` must match a CHT place id. If `place_id` is not included in the OpenMRS addresses, the mediator will attempt to match by the place name. This is not as reliable since the name must match exactly, and changes to either the CHT or OpenMRS will 

Author: removed all references to incoming patients

Patients that do not have an address or otherwise cannot be assigned a parent in CHT will be queryable in the FHIR Server and linked to OpenMRS patients, but will not be sent to CHT.

> After setting up the forms, test that it works in the test environment by creating a patient in OpenMRS.

Author: In the OpenMRS Utils postman collection there is OpenMRS Create Patient, but for this doc, removed all references to incoming patients

Log in to OpenHIM and view the transaction log. You should see:
1. After a maximum of 1 minute, a polling request from the OpenMRS mediator. The results from OpenMRS should include the newly created patient.
1. A request to the patient creation form.
1. A request from CHT to the `patient_ids` endpoint in the mediator
1. A request to the FHIR server updating the patient with the corresponding id from CHT.
1. A request to OpenMRS with the CHT id.

> If all the above look OK, you should now be able to see the patient in CHT.

Author: removed all references to incoming patients
